File: server/src/models/schedule.ts
import { DataTypes, Model, Optional } from 'sequelize';
import sequelize from '../config/database';
import User from './user';
import Workshift from './workshift';

interface ScheduleAttributes {
   id: number;
   userId: string;
   shiftId: number;
   startDate: string;
   endDate?: string | null;
}

interface ScheduleCreationAttributes extends Optional<ScheduleAttributes, 'id'> { }

class Schedule extends Model<ScheduleAttributes, ScheduleCreationAttributes> implements ScheduleAttributes {
   public declare id: number;
   public declare userId: string;
   public declare shiftId: number;
   public declare startDate: string;
   public declare endDate?: string | null;

   // Association mixins
   public declare Shift?: Workshift;
   public declare User?: User;

   public readonly createdAt!: Date;
   public readonly updatedAt!: Date;
}

Schedule.init({
   id: {
      type: DataTypes.INTEGER.UNSIGNED,
      autoIncrement: true,
      primaryKey: true,
   },
   userId: { type: DataTypes.UUID, allowNull: false },
   shiftId: { type: DataTypes.INTEGER.UNSIGNED, allowNull: false },
   startDate: { type: DataTypes.DATEONLY, allowNull: false },
   endDate: { type: DataTypes.DATEONLY, allowNull: true },
}, {
   sequelize,
   tableName: 'schedules',
   underscored: true,
   timestamps: true,
});

// Setup Associations (defined centrally in models/index.ts)
Schedule.belongsTo(User, { foreignKey: 'userId', as: 'User' });

export default Schedule;

File: client/lib/src/models/schedule.dart
import 'workshift.dart';
import 'user.dart';

class Schedule {
  final int id;
  final String userId;
  final int shiftId;
  final DateTime startDate;
  final DateTime? endDate;
  final Workshift? shift;
  final User? user;

  Schedule({
    required this.id,
    required this.userId,
    required this.shiftId,
    required this.startDate,
    this.endDate,
    this.shift,
    this.user,
  });

  factory Schedule.fromJson(Map<String, dynamic> json) {
    return Schedule(
      id: json['id'],
      userId: json['userId'],
      shiftId: json['shiftId'],
      startDate: DateTime.parse(json['startDate']),
      endDate: json['endDate'] != null ? DateTime.parse(json['endDate']) : null,
      shift: json['Shift'] != null ? Workshift.fromJson(json['Shift']) : null,
      user: json['User'] != null ? User.fromJson(json['User']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'shift_id': shiftId,
      'start_date': startDate.toIso8601String().substring(0, 10), // YYYY-MM-DD
      'end_date': endDate?.toIso8601String().substring(0, 10),
    };
  }
}
